pkg/storegateway/indexheader/lazy_binary_reader_test.go

```diff
@@ -294,6 +297,63 @@ func testLazyBinaryReader(t *testing.T, bkt objstore.BucketReader, dir string, i
 		return NewStreamBinaryReader(ctx, logger, bkt, dir, id, 3, NewStreamBinaryReaderMetrics(nil), Config{})
 	}
 
-	reader, err := NewLazyBinaryReader(ctx, factory, logger, bkt, dir, id, NewLazyBinaryReaderMetrics(nil), nil)
+	reader, err := NewLazyBinaryReader(ctx, factory, logger, bkt, dir, id, NewLazyBinaryReaderMetrics(nil), nil, gate.NewNoop())
 	test(t, reader, err)
 }
+
+// Tests if LazyBinaryReader blocks concurrent loads such that it doesn't pass the configured maximum
+func TestLazyBinaryReader_ShouldBlockMaxConcurrency(t *testing.T) {
+	ctx := context.Background()
+
+	tmpDir := filepath.Join(t.TempDir(), "test-indexheader")
+	bkt, err := filesystem.NewBucket(filepath.Join(tmpDir, "bkt"))
+	require.NoError(t, err)
+	t.Cleanup(func() { require.NoError(t, bkt.Close()) })
+
+	// Create block.
+	blockID, err := block.CreateBlock(ctx, tmpDir, []labels.Labels{
+		labels.FromStrings("a", "1"),
+		labels.FromStrings("a", "2"),
+		labels.FromStrings("a", "3"),
+	}, 100, 0, 1000, labels.FromStrings("ext1", "1"))
+	require.NoError(t, err)
+	require.NoError(t, block.Upload(ctx, log.NewNopLogger(), bkt, filepath.Join(tmpDir, blockID.String()), nil))
+
+	logger := log.NewNopLogger()
+	var total int
+	var inflight int
+	factory := func() (Reader, error) {
+		time.Sleep(3 * time.Second)
```

Comment:
Instead of relying on `Sleep()` to block long enough here, maybe we could use [`sync.Cond`](https://pkg.go.dev/sync#Cond) to be a little less racey?

Reply:
if i use `sync.Cond`, should i still have the unlock condition to be 3 seconds or some other condition?

Reply:
Hmm, I actually think the `sync.Cond` suggestion was a mistake.

The `time.Sleep()` and assertion on number of executing `factory` functions is the right path.

Example using a max concurrency of `1`:

```go
running := atomic.NewInt32(0)

factory := func() (Reader, error) {
    val := running.Inc()
    require.Equal(1, val)
    
    time.Sleep(time.Second)
    
    val := running.Dec()
    require.Equal(0, val)
    
    return nil, errors.New("oh no!")
}
```

The `sync.Cond()` might be useful to ensure each goroutine starts at exactly the same time instead of starting them one-by-one in a loop but it's not required.